### PR TITLE
RGRIDT-1027: [Maps] Leaflet Integration - Set directions [Part 2 - Directions Logic + CSS updates]

### PR DESCRIPTION
**Pull request by tmbp95**
This PR is for RGRIDT-1027: [Maps] Leaflet Integration - Set directions

### What was happening
* Leaflet Integration (parity feature to the existing Google Maps setDirections API)
* The wanted directions are shown on the Map when using "Provider" = Leaflet.

### What was done
* Implemented the Direction.ts feature for the LeafletProvider
* Added a new method setPlugin inside the IDirections.ts, LeafletProvider/Features/Directions.ts, and that same method (but not implemented) inside the GoogleProvider/Features/Directions.ts because the Interface demands it.
* Added the method LoadPlugin into the MapAPI.Directions.ts which will enable the new API method
* Added new ErrorCodes to the Map
* Fixed and added more CSS to replace the marker-icons and marker-tooltips from the waypoint icons

### Checklist
* [x] tested locally
* [x] documented the code
* [x] clean all warnings and errors of eslint
* [x] requires changes in OutSystems (if so, provide a module with changes) - **RGRIDT-1027**
* [ ] requires new sample page in OutSystems (if so, provide a module with changes)

### CODE REVIEW

**Review by JoaoFerreira-FrontEnd — APPROVED** [review]

File: code/src/LeafletProvider/Features/Directions.ts [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+                returnigMessage.isSuccess = true;
+            } else {
+                returnigMessage.code =
+                    OSFramework.Enum.ErrorCodes.API_FailedLoadingPlugin;
+            }
+            return returnigMessage;
+        }
+
+        public setRoute(
+            originRoute: string,
```
Comment (glundgren93) [inline code/src/LeafletProvider/Features/Directions.ts:339]: Can we put all of these parameters into a object/type?

File: code/src/LeafletProvider/Features/Directions.ts [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+            destinationRoute: string,
+            travelMode: string,
+            waypoints: string,
+            optimizeWaypoints: boolean,
+            avoidTolls: boolean,
+            avoidHighways: boolean,
+            avoidFerries: boolean
+        ): Promise<OSFramework.OSStructures.ReturnMessage> {
+            let returningMessage = new OSFramework.OSStructures.ReturnMessage();
+            this._currentDistance = NaN;
```
[the remark this answers is not in the document]
Comment (tmbp95) [reply code/src/LeafletProvider/Features/Directions.ts:361]: done!

File: code/src/LeafletProvider/Features/Directions.ts [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+            this.setState(false);
+            const returnigMessage =
+                new OSFramework.OSStructures.ReturnMessage();
+            if (!this._hasDirectionsRenderer()) {
+                returnigMessage.isSuccess = true;
+            } else {
+                returnigMessage.code =
+                    OSFramework.Enum.ErrorCodes.API_FailedRemoveDirections;
+            }
+            return returnigMessage;
```
Comment (rugoncalves) [inline code/src/LeafletProvider/Features/Directions.ts:320]: typo in variable name?

File: code/src/LeafletProvider/Features/Directions.ts [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+        }
+
+        public setPlugin(
+            provider: Constants.Directions.Provider | string,
+            apiKey: string
+        ): OSFramework.OSStructures.ReturnMessage {
+            if (this._directionsRenderer !== undefined) {
+                this.removeRoute();
+            }
```
Comment (rugoncalves) [inline code/src/LeafletProvider/Features/Directions.ts:329]: não temos de fazer unload do plugin anyhow? just confirming!

**Review by rugoncalves — CHANGES_REQUESTED** [review]

File: code/src/GoogleProvider/Features/Directions.ts [hunk 5]
```diff
@@ -66,33 +66,37 @@ namespace GoogleProvider.Feature {
 
             return legs;
         }
-        public getTotalDistanceFromDirection(): number {
-            // If the Map has the directions disabled return 0 (meters)
-            if (this._isEnabled === false) return 0;
+        public getTotalDistanceFromDirection(): Promise<number> {
+            return new Promise((resolve) => {
+                // If the Map has the directions disabled return 0 (meters)
+                if (this._isEnabled === false) return 0;
```
[the remark this answers is not in the document]
Comment (tmbp95) [reply code/src/GoogleProvider/Features/Directions.ts:72]: Thank you! Fixed it and added a better comment for the "If the Map has the directions disabled return 0 (meters)"

File: code/src/GoogleProvider/Features/Directions.ts [hunk 6]
```diff
@@ -106,34 +110,47 @@ namespace GoogleProvider.Feature {
...
+        public setPlugin(
+            // eslint-disable-next-line @typescript-eslint/no-unused-vars
+            providerName: string,
+            // eslint-disable-next-line @typescript-eslint/no-unused-vars
+            apiKey: string
+        ): OSFramework.OSStructures.ReturnMessage {
+            OSFramework.Helper.ThrowError(
+                this._map,
+                OSFramework.Enum.ErrorCodes.GEN_NoPluginDirectionsNeeded
+            );
```
[the remark this answers is not in the document]
Comment (tmbp95) [reply code/src/GoogleProvider/Features/Directions.ts:126]: We can discuss that with the team...

File: code/src/MapAPI/Directions.ts [hunk 7]
```diff
@@ -13,7 +13,9 @@ namespace MapAPI.Directions {
      * Function that will calculate the total distance of the direction that is rendered on the Map.
      * @param mapId Id of the Map where the total distance of the direction will be calculated.
      */
-    export function GetTotalDistanceFromDirection(mapId: string): number {
+    export function GetTotalDistanceFromDirection(
+        mapId: string
+    ): Promise<number> {
```
[the remark this answers is not in the document]
Comment (tmbp95) [reply code/src/MapAPI/Directions.ts:18]: We are already going to create a lot of new breaking changes...

File: code/src/LeafletProvider/Features/Directions.ts [hunk 8]
```diff
@@ -0,0 +1,417 @@
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+namespace LeafletProvider.Feature {
+    enum ResolveType {
```
[the remark this answers is not in the document]
Comment (tmbp95) [reply code/src/LeafletProvider/Features/Directions.ts:3]: thank you!

File: code/src/LeafletProvider/Features/Directions.ts [hunk 9]
```diff
@@ -0,0 +1,417 @@
...
+            this._currentDuration = e.routes[0].summary.totalTime;
+            // If the resolve exists, means that this handler was invoked by the methods getDistance or getDuration as it has been passed via "bind"
+            // Otherwise, it means that the handler was invoked by the method setRoute on the creation of the route
+            if (resolve !== undefined) {
+                if (resolveType === ResolveType.GetDistance) {
+                    // Make sure to resolve with the callback "resolve" and passing the currentDistance as an argument
+                    resolve(this._currentDistance);
+                    // invoked by the getDistance method -> as soon as the routesFoundHandler gets executed and after getting the currentDistance and currentDuration we can remove the listener for the getDistance.
+                    this._directionsRenderer.off(
+                        'routesfound',
```
[the remark this answers is not in the document]
Comment (tmbp95) [reply code/src/LeafletProvider/Features/Directions.ts:144]: done!

File: code/src/LeafletProvider/Features/Directions.ts [hunk 10]
```diff
@@ -0,0 +1,417 @@
...
+         * Can be used by the GetDistance or GetDuration methods and therefore will have specific attributes binded
+         * @param resolve Will be the resolve method from the Promise
+         * @param isDistance In case of a GetDistance this Boolean should be True. Otherwise, it should be False.
+         * @param e Event that got triggered (L.Routing.RoutingResultEvent) and respective structure
+         */
+        private _routesFoundHandler(
+            resolve?: OSFramework.Callbacks.Generic,
+            resolveType?: ResolveType,
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+            e?: any
```
[the remark this answers is not in the document]
Comment (tmbp95) [reply code/src/LeafletProvider/Features/Directions.ts:130]: Tried to find the L.Routing.RoutingResultEvent but I couldn't find it...

File: code/src/LeafletProvider/Features/Directions.ts [hunk 11]
```diff
@@ -0,0 +1,417 @@
...
+                        this._bindDuration
+                    );
+                }
+            } else {
+                if (resolveType === ResolveType.SetRoute) {
+                    // invoked by the setRoute method -> as soon as the routesFoundHandler gets executed and after getting the currentDistance and currentDuration we can remove the listener for the setRoute.
+                    this._directionsRenderer.off(
+                        'routesfound',
+                        this._bindSetRoute
+                    );
```
[the remark this answers is not in the document]
Comment (tmbp95) [reply code/src/LeafletProvider/Features/Directions.ts:162]: done

File: code/src/LeafletProvider/Features/Directions.ts [hunk 12]
```diff
@@ -0,0 +1,417 @@
...
+        }
+
+        /**
+         * Sets the avoidance criteria (avoidTolls, avoidHighways, avoidFerries) on the directionsRenderer routing service
+         */
+        private _setExcludes(
+            dirExclude: OSFramework.OSStructures.Directions.ExcludeCriteria
+        ): boolean {
+            // If the Map has no directionsRenderer, return false.
+            if (!this._hasDirectionsRenderer()) return false;
```
[the remark this answers is not in the document]
Comment (tmbp95) [reply code/src/LeafletProvider/Features/Directions.ts:174]: done

File: code/src/LeafletProvider/Features/Directions.ts [hunk 13]
```diff
@@ -0,0 +1,417 @@
...
+        }
+
+        /**
+         * Sets the travelMode on the directionsRenderer routing service
+         */
+        private _setTravelMode(travelMode: string): boolean {
+            // If the Map has no directionsRenderer, return false.
+            // If the provider of the routing service doesn't have the chosen travelMode option, return false
+            if (
+                !this._hasDirectionsRenderer() ||
```
[the remark this answers is not in the document]
Comment (tmbp95) [reply code/src/LeafletProvider/Features/Directions.ts:193]: done

File: code/src/LeafletProvider/Features/Directions.ts [hunk 14]
```diff
@@ -0,0 +1,417 @@
...
+        }
+
+        /** This method is not implemented for the LeafletProvider Directions API */
+        public getLegsFromDirection(): Array<OSFramework.OSStructures.Directions.DirectionLegs> {
+            throw new Error('Method not implemented.');
+        }
+
+        public getTotalDistanceFromDirection(): Promise<number> {
+            // Validate if the directionsRenderer exists
+            if (!this._hasDirectionsRenderer()) {
```
[the remark this answers is not in the document]
Comment (tmbp95) [reply code/src/LeafletProvider/Features/Directions.ts:256]: done

File: code/src/LeafletProvider/Features/Directions.ts (hunk 3 above)
Comment (tmbp95) [reply code/src/LeafletProvider/Features/Directions.ts:320]: yep :p thank you!

File: code/src/LeafletProvider/Features/Directions.ts (hunk 4 above)
Comment (tmbp95) [reply code/src/LeafletProvider/Features/Directions.ts:329]: What would be the point of unloading it?

File: code/src/GoogleProvider/Features/Directions.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (rugoncalves) [reply code/src/GoogleProvider/Features/Directions.ts:126]: Just asking :P

File: code/src/MapAPI/Directions.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (rugoncalves) [reply code/src/MapAPI/Directions.ts:18]: True, let's list this in the change log - please!

File: code/src/OSFramework/Feature/IDirections.ts [hunk 15]
```diff
@@ -5,34 +5,36 @@ namespace OSFramework.Feature {
...
-        /** Get Total Distance from Direction (sum in meters of all the distances from all legs) */
-        getTotalDistanceFromDirection(): number;
-        /** Get Total Duration from Direction (sum in seconds of all the durations from all legs) */
-        getTotalDurationFromDirection(): number;
+        /** Get Total Distance from Direction (sum in meters of all the distances from all legs).
+         */
+        getTotalDistanceFromDirection(): Promise<number>;
+        /** Get Total Duration from Direction (sum in seconds of all the durations from all legs)
+         */
+        getTotalDurationFromDirection(): Promise<number>;
```
[the remark this answers is not in the document]
Comment (rugoncalves) [reply code/src/OSFramework/Feature/IDirections.ts:13]: We'd need is to validate if typeof "number" or not... ugly - but it itches a bit doing that for GM when it's so straight forward. Anyone else thoughts? @bmartinho ?

File: code/src/LeafletProvider/Features/Directions.ts (hunk 4 above)
Comment (rugoncalves) [reply code/src/LeafletProvider/Features/Directions.ts:326]: what happens when we start loading multiple plugins simultaneously? Sounds strange. Unloading would be to call it's destroy or so?

File: code/src/MapAPI/Directions.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (rugoncalves) [reply code/src/MapAPI/Directions.ts:18]: (close this comment, once done)

File: code/src/LeafletProvider/Features/Directions.ts [hunk 16]
```diff
@@ -0,0 +1,417 @@
...
+                    // Make sure to resolve with the callback "resolve" and passing the currentDuration as an argument
+                    resolve(this._currentDuration);
+                    // invoked by the getDuration method -> as soon as the routesFoundHandler gets executed and after getting the currentDistance and currentDuration we can later remove the listener for the getDuration.
+                    bind = this._bindDuration;
+                }
+            } else {
+                if (resolveType === ResolveType.SetRoute) {
+                    // invoked by the setRoute method -> as soon as the routesFoundHandler gets executed and after getting the currentDistance and currentDuration we can later remove the listener for the setRoute.
+                    bind = this._bindSetRoute;
+                }
```
Comment (rugoncalves) [inline code/src/LeafletProvider/Features/Directions.ts:155]: shouldn't we resolve the promise never the less?

File: code/src/OSFramework/Feature/IDirections.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (tmbp95) [reply code/src/OSFramework/Feature/IDirections.ts:13]: Is that a big problem having a Promise?

File: code/src/LeafletProvider/Features/Directions.ts (hunk 4 above)
Comment (tmbp95) [reply code/src/LeafletProvider/Features/Directions.ts:326]: If multiple loadPlugin actions get called simultaneously, that wouldn't be the best approach to do directions right? 
In my opinion if the developer wants to loadPlugin multiple times, that would mean that he must not want to use the previously loaded one anymore... Also, if the developers wants to set 2 plugins and then use 2 setDirections with one for the 1st plugin and the other for the 2nd plugin... That wouldn't be possible without having a new parameter for the setDirections...

File: code/src/LeafletProvider/Features/Directions.ts (hunk 16 above)
Comment (tmbp95) [reply code/src/LeafletProvider/Features/Directions.ts:155]: We don't have a promise for SetRoute. We just want to update the this.currentDistance and this.currentDuration variables

File: code/src/MapAPI/Directions.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (tmbp95) [reply code/src/MapAPI/Directions.ts:18]: added!

**Review by rugoncalves — DISMISSED** [review]

**Review by rugoncalves — APPROVED** [review]

**Review by JoaoFerreira-FrontEnd — APPROVED** [review]

**Pull request merged**